example/convolution2D/src/convolution2D.cpp

```diff
@@ -0,0 +1,374 @@
+                    auto const exactCol = static_cast<int32_t>(col) - filterWidth / 2 + fCol;
+                    if(exactRow >= 0 && exactRow < matrixHeight && exactCol >= 0 && exactCol < matrixWidth)
+                    {
+                        pValue += filter[fRow * filterWidthAllocated + fCol]
+                                  * input[exactRow * intputWidthAllocated + exactCol];
+                    }
+                }
+                output[row * matrixWidth + col] = pValue;
+            }
+        }
+    }
+};
+
+/**
+ * @brief ConvolutionKernel2DSharedMemory struct. The kernel for 2D Convolutional Filter, uses
+ tiling method. Tiles of matrix are kept in the shared memory. Block
+ dimensions are equal to tile dimensions.
+ */
+struct ConvolutionKernel2DSharedMemory
+{
+    /**
+    @tparam TAcc Accelerator type
+    @tparam TElem The input-matrix and filter-matrix element type
+    @param acc Accelerator
+    @param input Input matrix
+    @param output Output matrix
+    @param matrixWidth Input matrix width
+    @param matrixHeight Input matrix height
+    @param filter Filter-matrix
+    @param filterWidth Filter-matrix width
+    @param intputWidthAllocated Input-matrix width allocated (possibly larger than normal width due to paddding
+    @param filterWidthAllocated Filter-matrix width allocated (possibly larger than normal width due to paddding
+        */
+    template<typename TAcc, typename TElem>
+    ALPAKA_FN_ACC auto operator()(
+        TAcc const& acc,
+        TElem const* const input,
+        TElem* output,
+        int32_t const matrixWidth,
+        int32_t const matrixHeight,
+        TElem const* const filter,
+        int32_t const filterWidth,
+        int32_t const intputWidthAllocated,
+        int32_t const filterWidthAllocated) const -> void
+    {
+        auto const [row, col] = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc);
+        // Get extents(dimensions)
+        auto const gridBlockExtent = alpaka::getWorkDiv<alpaka::Grid, alpaka::Blocks>(acc);
+        auto const blockThreadExtent = alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc);
+        // Get indexes
+        auto const blockThreadIdx = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc);
+        auto const blockThreadIdx1D = alpaka::mapIdx<1u>(blockThreadIdx, blockThreadExtent)[0u];
+        // Get elements from 2-element arrays
+        auto const [blockThreadExtentY, blockThreadExtentX] = blockThreadExtent;
+        auto const [blockThreadY, blockThreadX] = blockThreadIdx;
+        auto const [gridBlockExtentY, gridBlockExtentX] = gridBlockExtent;
+        // Allocate shared memory
+        auto* const sharedN = alpaka::getDynSharedMem<TElem>(acc);
+        // Fill shared memory of device so that tile items are accessed from shared memory
+        if(row < matrixHeight && col < matrixWidth && blockThreadIdx1D < blockThreadExtent.prod())
```

Comment:
On the host side you use `getValidWorkDiv` This means you will have one thread for some alpaka accelerators.
I know you wrote that the block size must be equal to the tile size but you do not enforce  it e.g. with an `ALPAKA_VERIFY`
If you have only one thread in the block you can simply iterate over the shared memory to fill it.

Reply:
Yes but if there is one thread it means it is not a GPU; so we dont know which level of memory is used ?